docs/src/main/asciidoc/cdi-integration.adoc

```diff
@@ -0,0 +1,606 @@
+[source,java]
+----
+@BuildStep
+ObserverRegistrarBuildItem syntheticObserver() {
+    return new ObserverRegistrarBuildItem(new ObserverRegistrar() {
+       @Override
+       public void register(RegistrationContext context) {
+          // NOTE: You can't produce a build item from within this callback!
+          context.configure().observedType(String.class) <1>
+                             .notify(mc -> {
+                               // do some gizmo bytecode generation...
+                             }).done();
+            }
+        }));
+}
+----
+<1> The output of a `ObserverConfigurator` is recorded as bytecode. Therefore, there are some limitations in how a synthetic observer is invoked at runtime. Currently, you must generate the bytecode of the method body directly.
+
+=== _Solution 2_: `ObserverRegistrationPhaseBuildItem`
+
+If a build step *needs to produce other build items during the registration* it should use the `ObserverRegistrationPhaseBuildItem`.
+The reason is that an injected `BuildProducer` is only valid during a `@BuildStep` method invocation.
+
+IMPORTANT: A build step that consumes the `ObserverRegistrationPhaseBuildItem` should always produce an `ObserverConfiguratorBuildItem` or at least inject a `BuildProducer` for this build item, otherwise it could be ignored or processed at the wrong time (e.g. after the correct CDI bootstrap phase).
+
+.`ObserverRegistrationPhaseBuildItem` Example
+[source,java]
+----
+@BuildStep
+void syntheticObserver(ObserverRegistrationPhaseBuildItem observerRegistrationPhase,
+            BuildProducer<MyBuildItem> myBuildItem, 
+            BuildProducer<ObserverConfiguratorBuildItem> observerConfigurators) {
+   observerConfigurators.produce(new ObserverConfiguratorBuildItem(observerRegistrationPhase.getContext().configure().observedType(String.class)
+                             .notify(mc -> {
+                               // do some gizmo bytecode generation...
+                             }).done();
+   myBuildItem.produce(new MyBuildItem());
+}
+----
+
+[[generated_beans]]
+== Use Case - I Have a Generated Bean Class
+
+No problem.
+You can generate the bytecode of a bean class manually and then all you need to do is to produce a `GeneratedBeanBuildItem` instead of `GeneratedClassBuildItem`.
+
+.`GeneratedBeanBuildItem` Example
+[source,java]
+----
+@BuildStep
+GeneratedBeanBuildItem generatedBean(BuildProducer<GeneratedBeanBuildItem> generatedBeans) {
+    ClassOutput beansClassOutput = new GeneratedBeanGizmoAdaptor(generatedBeans); <1>
+    ClassCreator beanClassCreator = ClassCreator.builder().classOutput(beansClassOutput)
+                .className("org.acme.MyBean")
+                .build();
+    beanClassCreator.addAnnotation(Singleton.class);
+    beanClassCreator.close(); <2>
+}
+----
+<1> A convenient `io.quarkus.arc.deployment.GeneratedBeanGizmoAdaptor` can be used to produce a `GeneratedBeanBuildItem` from Gizmo constructs.
```

Comment:
I would rephrase as: Using `io.quarkus.arc.deployment.GeneratedBeanGizmoAdaptor` makes it easy to produce `GeneratedBeanBuildItem` from Gizmo constructs